src/els_diagnostics_server.erl

```diff
@@ -0,0 +1,168 @@
+-record(state, { server_name :: atom()
+               , uri :: uri()
+               , diagnostics :: term()
+               }).
+
+%%==============================================================================
+%% Type Definitions
+%%==============================================================================
+-type state() :: #state{}.
+
+%%==============================================================================
+%% API
+%%==============================================================================
+-spec start_link({atom(), uri()}, term()) -> {ok, pid()}.
+start_link({Id, Uri}, _Opts) ->
+  {ok, _} = gen_server:start_link({local, Id},
+                                  ?MODULE,
+                                  {Id, Uri},
+                                  []).
+
+%%==============================================================================
+%% gen_server callbacks
+%%==============================================================================
+-spec init({atom(), uri()}) -> {ok, state()}.
+init({Id, Uri}) ->
+  State = #state{ server_name = Id
+                , uri = Uri
+                },
+  {ok, State}.
+
+-spec handle_call(any(), any(), state()) -> {reply, any(), state()}.
+handle_call(get_diagnostics, _From, #state{diagnostics = Diagnostics} = S) ->
+  {reply, Diagnostics, S};
+handle_call(UnknownMessage, _From, S) ->
+  lager:info("Unknown handle_call! message: ~p", [UnknownMessage]),
+  {reply, ok, S}.
+
+-spec handle_cast(any(), state()) -> {noreply, state()}.
+handle_cast(run, #state{uri = Uri} = S) ->
+  Diagnostics =
+    lists:foldl(
+      fun(Mod, AccumulatedDiagnostics0) ->
+          Diagnostics = diagnostic(Mod, Uri),
+          AccumulatedDiagnostics = AccumulatedDiagnostics0 ++ Diagnostics,
+          send_notification(Uri, AccumulatedDiagnostics),
+          AccumulatedDiagnostics
+      end, [], [ els_compiler_diagnostics
+               , els_elvis_diagnostics
+               , els_dialyzer_diagnostics
+               ]),
+  {noreply, S#state{diagnostics = Diagnostics}};
+handle_cast(UnknownMessage, S) ->
+  lager:info("Unknown handle_cast! message: ~p", [UnknownMessage]),
+  {noreply, S}.
+
+-spec handle_info(any(), state()) -> {noreply, state()}.
+handle_info({'EXIT', _, normal}, S) ->
+  {noreply, S};
+handle_info(UnknownMessage, S) ->
+  lager:info("Unknown handle_info! message: ~p", [UnknownMessage]),
+  {noreply, S}.
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+-spec on_save(uri()) -> ok.
+on_save(Uri) ->
+  els_diagnostics_sup:on_save(Uri),
+  Id = binary_to_atom(Uri, utf8),
+  ok = gen_server:cast(Id, run).
+
+-spec on_open(uri()) -> ok.
+on_open(Uri) ->
+  els_diagnostics_sup:on_open(Uri),
+  Id = binary_to_atom(Uri, utf8),
+  case gen_server:call(Id, get_diagnostics, 100) of
```

Comment:
Are these diagnostics sorted by location?  The first 100 would make more sense than a random set.

Reply:
The 100 is the timeout in ms for this call. I found some cases where we got a timeout on this call and I wanted to shorten it as much as what I deemed reasonable.

For each uri we spin up a new gen_server so all gen_server are isolated from eachother and you can quickly jump between files, save or open and get the diagnostics async. (For example openening multiple files starts to fire up a bunch of gen_server and starts creating diagnostics for each and fire off a bunch of Publish Diagnostics messages.

Reply:
I would remove the `get_diagnostics` feature, kill the diagnostics process for each `Uri` when the file is closed and just run the diagnostics again when it is opened. Therefore avoiding this logic completely.

Reply:
It feels kinda dirty to add that call to `els_client` which is the only place that handles the `did_close`. Re-running diagnostics on `didOpen` every time in particular when Dialyzer is enabled (or a large file is opened) can be quite expensive. Reporting what's already there can save a lot of overhead and time. 

Can you elaborate why you'd want to get rid of this?

Reply:
Elvis is very cheap, Dialyzer is not be, but optimizing for the special case of people opening and closing the same file all the time is not the way to go IMO.

Reply:
Fair enough :+1: